src/unauthenticated_stream/mod.rs

```diff
@@ -0,0 +1,224 @@
+    key: Key,
+}
+
+impl EncryptingKey {
+    /// Create a new encrypting key.
+    ///
+    /// `key_bytes` must be exactly `algorithm.key_len` bytes long.
+    ///
+    /// C analogs: `EVP_CIPHER_CTX_init`.
+    ///
+    /// Go analog:
+    ///   [`crypto.aes.NewCipher`](https://golang.org/pkg/crypto/aes/#NewCipher)
+    #[inline]
+    pub fn new(algorithm: &'static Algorithm, key_bytes: &[u8])
+               -> Result<EncryptingKey, error::Unspecified> {
+        Ok(EncryptingKey {
+            key: Key::new(algorithm, key_bytes)?,
+        })
+    }
+
+    /// The key's unauthenticated stream cipher algorithm.
+    ///
+    /// C analog: `EVP_CIPHER_CTX.cipher`
+    #[inline(always)]
+    pub fn algorithm(&self) -> &'static Algorithm { self.key.algorithm() }
+}
+
+/// Encrypts the input data in place.
+pub fn encrypt_in_place(key: &EncryptingKey, nonce: &[u8], in_out: &mut [u8])
+                        -> Result<usize, error::Unspecified> {
+    let nonce = slice_as_array_ref!(nonce, NONCE_LEN)?;
+    check_per_nonce_max_bytes(key.key.algorithm, in_out.len())?;
+    (key.key.algorithm.xor_keystream)(&key.key.ctx_buf, nonce, in_out)?;
+    Ok(in_out.len())
+}
+
+/// `DecryptingKey` and `EncryptingKey` are type-safety wrappers around `Key`,
+/// which does all the actual work via the C stream cipher interface.
+///
+/// C analog: `EVP_CIPHER_CTX`
+struct Key {
+    ctx_buf: [u64; KEY_CTX_BUF_ELEMS],
+    algorithm: &'static Algorithm,
+}
+
+const KEY_CTX_BUF_ELEMS: usize = (KEY_CTX_BUF_LEN + 7) / 8;
+
+const KEY_CTX_BUF_LEN: usize = chacha::KEY_LEN_IN_BYTES;
+
+impl Key {
+    fn new(algorithm: &'static Algorithm, key_bytes: &[u8]) -> Result<Self, error::Unspecified> {
+        if key_bytes.len() != algorithm.key_len() {
+            return Err(error::Unspecified);
+        }
+
+        let mut r = Key {
+            algorithm,
+            ctx_buf: [0; KEY_CTX_BUF_ELEMS],
+        };
+
+        init::init_once();
+        {
+            let ctx_buf_bytes = polyfill::slice::u64_as_u8_mut(&mut r.ctx_buf);
+            (r.algorithm.init)(ctx_buf_bytes, key_bytes)?;
+        }
+
+        Ok(r)
+    }
+
+    /// The key's unauthenticated stream cipher algorithm.
+    #[inline(always)]
+    fn algorithm(&self) -> &'static Algorithm { self.algorithm }
+}
+
+/// An unauthenticated stream cipher Algorithm.
+///
+/// C analog: `EVP_CIPHER`
+///
+/// Go analog:
+///     [`crypto.cipher.Stream`](https://golang.org/pkg/crypto/cipher/#Stream)
+pub struct Algorithm {
+    init: fn(ctx_buf: &mut [u8], key: &[u8]) -> Result<(), error::Unspecified>,
+
+    xor_keystream: fn(ctx: &[u64; KEY_CTX_BUF_ELEMS], nonce: &[u8; NONCE_LEN],
+                      in_out: &mut [u8]) -> Result<(), error::Unspecified>,
+
+    key_len: usize,
+    id: AlgorithmID,
+
+    /// Use `max_input_len!()` to initialize this.
+    // TODO: Make this `usize`.
+    max_input_len: u64,
+}
+
+/// TODO: Make this a `const fn` when those become stable.
+macro_rules! max_input_len {
```

Comment:
I think we can/should avoid duplicating this.

Reply:
Done.